File: cozkazan/lib/services/duygu_analizi_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../constants.dart';

class DuyguAnalizi {
  final String id;
  final String kullaniciId;
  final String kullaniciTipi;
  final String mesaj;
  final double duyguSkoru;
  final String duyguKategori;
  final List<String> anaDuygular;
  final TonAnalizi tonAnalizi;
  final String dil;
  final DateTime tarih;
  final String yanitToni;
  final bool yanitAyarlandi;

  DuyguAnalizi({
    required this.id,
    required this.kullaniciId,
    required this.kullaniciTipi,
    required this.mesaj,
    required this.duyguSkoru,
    required this.duyguKategori,
    required this.anaDuygular,
    required this.tonAnalizi,
    required this.dil,
    required this.tarih,
    required this.yanitToni,
    required this.yanitAyarlandi,
  });

  factory DuyguAnalizi.fromJson(Map<String, dynamic> json) {
    return DuyguAnalizi(
      id: json['_id'],
      kullaniciId: json['kullaniciId'],
      kullaniciTipi: json['kullaniciTipi'],
      mesaj: json['mesaj'],
      duyguSkoru: json['duyguSkoru'].toDouble(),
      duyguKategori: json['duyguKategori'],
      anaDuygular: List<String>.from(json['anaDuygular']),
      tonAnalizi: TonAnalizi.fromJson(json['tonAnalizi']),
      dil: json['dil'],
      tarih: DateTime.parse(json['tarih']),
      yanitToni: json['yanitToni'],
      yanitAyarlandi: json['yanitAyarlandi'],
    );
  }
}

class TonAnalizi {
  final double resmiyet;
  final double enerji;
  final double aciliyet;

  TonAnalizi({
    required this.resmiyet,
    required this.enerji,
    required this.aciliyet,
  });

  factory TonAnalizi.fromJson(Map<String, dynamic> json) {
    return TonAnalizi(
      resmiyet: json['resmiyet'].toDouble(),
      enerji: json['enerji'].toDouble(),
      aciliyet: json['aciliyet'].toDouble(),
    );
  }
}

class DuyguIstatistikleri {
  final int toplamMesaj;
  final double ortalamaDuyguSkoru;
  final Map<String, int> duyguKategorileri;
  final Map<String, int> enCokKullanilanDuygular;
  final TonAnalizi tonAnaliziOrtalamalari;

  DuyguIstatistikleri({
    required this.toplamMesaj,
    required this.ortalamaDuyguSkoru,
    required this.duyguKategorileri,
    required this.enCokKullanilanDuygular,
    required this.tonAnaliziOrtalamalari,
  });

  factory DuyguIstatistikleri.fromJson(Map<String, dynamic> json) {
    return DuyguIstatistikleri(
      toplamMesaj: json['toplamMesaj'],
      ortalamaDuyguSkoru: json['ortalamaDuyguSkoru'].toDouble(),
      duyguKategorileri: Map<String, int>.from(json['duyguKategorileri']),
      enCokKullanilanDuygular: Map<String, int>.from(json['enCokKullanilanDuygular']),
      tonAnaliziOrtalamalari: TonAnalizi.fromJson(json['tonAnaliziOrtalamalari']),
    );
  }
}

class DuyguAnaliziService {
  static const String baseUrl = 'http://91.99.187.116:3000';

  // Mesajın duygusal tonunu analiz et
  static Future<DuyguAnalizi> analizEt({
    required String mesaj,
    required String kullaniciId,
    required String kullaniciTipi,
    String dil = 'tr',
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/duygu-analizi/analiz'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'mesaj': mesaj,
          'kullaniciId': kullaniciId,
          'kullaniciTipi': kullaniciTipi,
          'dil': dil,
        }),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success'] == true) {
          return DuyguAnalizi.fromJson(data['data']);
        } else {
          throw Exception('Duygu analizi başarısız: ${data['message']}');
        }
      } else {
        throw Exception('Duygu analizi hatası: ${response.statusCode}');
      }
    } catch (e) {
      print('Duygu analizi servis hatası: $e');
      // Fallback: Basit duygu analizi
      return _getFallbackAnaliz(mesaj, kullaniciId, kullaniciTipi);
    }
  }

  // Yanıtı duygu analizine göre ayarla
  static Future<String> yanitiAyarla({
    required String yanit,
    required String duyguAnaliziId,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/duygu-analizi/yanit-ayarla'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'yanit': yanit,
          'duyguAnaliziId': duyguAnaliziId,
        }),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success'] == true) {
          return data['data']['ayarlanmisYanit'];
        } else {
          throw Exception('Yanıt ayarlama başarısız: ${data['message']}');
        }
      } else {
        throw Exception('Yanıt ayarlama hatası: ${response.statusCode}');
      }
    } catch (e) {
      print('Yanıt ayarlama servis hatası: $e');
      return yanit; // Fallback: Orijinal yanıtı döndür
    }
  }

  // Kullanıcının duygu geçmişini getir
  static Future<List<DuyguAnalizi>> kullaniciDuyguGecmisi({
    required String kullaniciId,
    required String kullaniciTipi,
    int gunSayisi = 7,
  }) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/duygu-analizi/gecmis/$kullaniciId?kullaniciTipi=$kullaniciTipi&gunSayisi=$gunSayisi'),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success'] == true) {
          return (data['data'] as List)
              .map((json) => DuyguAnalizi.fromJson(json))
              .toList();
        } else {
          throw Exception('Duygu geçmişi getirme başarısız: ${data['message']}');
        }
      } else {
        throw Exception('Duygu geçmişi getirme hatası: ${response.statusCode}');
      }
    } catch (e) {
      print('Duygu geçmişi servis hatası: $e');
      return [];
    }
  }

  // Duygu istatistiklerini getir
  static Future<DuyguIstatistikleri?> duyguIstatistikleri({
    required String kullaniciId,
    required String kullaniciTipi,
    int gunSayisi = 30,
  }) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/duygu-analizi/istatistikler/$kullaniciId?kullaniciTipi=$kullaniciTipi&gunSayisi=$gunSayisi'),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success'] == true) {
          return DuyguIstatistikleri.fromJson(data['data']);
        } else {
          throw Exception('İstatistik getirme başarısız: ${data['message']}');
        }
      } else {
        throw Exception('İstatistik getirme hatası: ${response.statusCode}');
      }
    } catch (e) {
      print('Duygu istatistikleri servis hatası: $e');
      return null;
    }
  }

  // Basit duygu analizi (fallback)
  static DuyguAnalizi _getFallbackAnaliz(String mesaj, String kullaniciId, String kullaniciTipi) {
    final lowerMesaj = mesaj.toLowerCase();
    double duyguSkoru = 0.0;
    List<String> anaDuygular = ['nötr'];

    // Basit kelime analizi
    final pozitifKelimeler = ['mutlu', 'sevinçli', 'harika', 'süper', 'güzel', 'iyi', 'başarılı'];
    final negatifKelimeler = ['üzgün', 'kızgın', 'kötü', 'berbat', 'yorgun', 'stresli', 'endişeli'];

    int pozitifSayisi = 0;
    int negatifSayisi = 0;

    for (final kelime in pozitifKelimeler) {
      if (lowerMesaj.contains(kelime)) {
        pozitifSayisi++;
        anaDuygular = ['mutlu'];
      }
    }

    for (final kelime in negatifKelimeler) {
      if (lowerMesaj.contains(kelime)) {
        negatifSayisi++;
        anaDuygular = ['üzgün'];
      }
    }

    if (pozitifSayisi > negatifSayisi) {
      duyguSkoru = 0.5;
    } else if (negatifSayisi > pozitifSayisi) {
      duyguSkoru = -0.5;
    }

    // Emoji analizi
    final pozitifEmojiler = ['😊', '😄', '😃', '😁', '😆', '😍', '🥰', '👍', '👏', '🎉', '✨', '🌟', '💪', '🔥', '❤️'];
    final negatifEmojiler = ['😢', '😭', '😤', '😠', '😡', '🤬', '😈', '👿', '💀', '☠️', '💩'];

    for (final emoji in pozitifEmojiler) {
      if (mesaj.contains(emoji)) {
        duyguSkoru += 0.3;
        anaDuygular = ['mutlu'];
      }
    }

    for (final emoji in negatifEmojiler) {
      if (mesaj.contains(emoji)) {
        duyguSkoru -= 0.3;
        anaDuygular = ['üzgün'];
      }
    }

    // Skoru sınırla
    duyguSkoru = duyguSkoru.clamp(-1.0, 1.0);

    // Duygu kategorisini belirle
    String duyguKategori;
    if (duyguSkoru <= -0.6) {
      duyguKategori = 'çok_negatif';
    } else if (duyguSkoru <= -0.2) {
      duyguKategori = 'negatif';
    } else if (duyguSkoru <= 0.2) {
      duyguKategori = 'nötr';
    } else if (duyguSkoru <= 0.6) {
      duyguKategori = 'pozitif';
    } else {
      duyguKategori = 'çok_pozitif';
    }

    // Ton analizi
    final tonAnalizi = TonAnalizi(
      resmiyet: 0.5, // Varsayılan orta
      enerji: duyguSkoru > 0 ? 0.7 : 0.3, // Pozitif ise yüksek enerji
      aciliyet: 0.3, // Varsayılan düşük
    );

    // Yanıt tonunu belirle
    String yanitToni;
    if (kullaniciTipi == 'child') {
      if (duyguSkoru < -0.3) {
        yanitToni = 'destekleyici';
      } else if (duyguSkoru < 0.2) {
        yanitToni = 'motivasyonel';
      } else {
        yanitToni = 'samimi';
      }
    } else {
      if (duyguSkoru < -0.2) {
        yanitToni = 'destekleyici';
      } else {
        yanitToni = 'eğitici';
      }
    }

    return DuyguAnalizi(
      id: DateTime.now().millisecondsSinceEpoch.toString(),
      kullaniciId: kullaniciId,
      kullaniciTipi: kullaniciTipi,
      mesaj: mesaj,
      duyguSkoru: duyguSkoru,
      duyguKategori: duyguKategori,
      anaDuygular: anaDuygular,
      tonAnalizi: tonAnalizi,
      dil: 'tr',
      tarih: DateTime.now(),
      yanitToni: yanitToni,
      yanitAyarlandi: false,
    );
  }

  // Duygu skoruna göre emoji önerisi
  static String getEmojiOnerisi(double duyguSkoru) {
    if (duyguSkoru > 0.5) {
      return '😊';
    } else if (duyguSkoru > 0.2) {
      return '🙂';
    } else if (duyguSkoru > -0.2) {
      return '😐';
    } else if (duyguSkoru > -0.5) {
      return '😔';
    } else {
      return '😢';
    }
  }

  // Duygu kategorisine göre renk
  static String getDuyguRengi(String duyguKategori) {
    switch (duyguKategori) {
      case 'çok_pozitif':
        return '#28a745'; // Yeşil
      case 'pozitif':
        return '#20c997'; // Açık yeşil
      case 'nötr':
        return '#6c757d'; // Gri
      case 'negatif':
        return '#fd7e14'; // Turuncu
      case 'çok_negatif':
        return '#dc3545'; // Kırmızı
      default:
        return '#6c757d';
    }
  }

  // Yanıt tonuna göre öneri
  static String getYanitToniOnerisi(String yanitToni) {
    switch (yanitToni) {
      case 'destekleyici':
        return 'Destekleyici ve anlayışlı bir ton kullanın';
      case 'motivasyonel':
        return 'Motivasyonel ve cesaretlendirici bir ton kullanın';
      case 'eğitici':
        return 'Eğitici ve bilgilendirici bir ton kullanın';
      case 'samimi':
        return 'Samimi ve arkadaşça bir ton kullanın';
      case 'profesyonel':
        return 'Profesyonel ve resmi bir ton kullanın';
      default:
        return 'Uygun bir ton seçin';
    }
  }
} 

File: backend/services/duyguAnaliziService.js
import DuyguAnalizi from '../models/DuyguAnalizi.js';

class DuyguAnaliziService {
  // Türkçe duygu kelimeleri sözlüğü
  static duyguKelimeleri = {
    pozitif: [
      'mutlu', 'sevinçli', 'heyecanlı', 'gururlu', 'başarılı', 'güvenli', 'huzurlu',
      'neşeli', 'coşkulu', 'enerjik', 'motivasyonlu', 'umutlu', 'içten', 'samimi',
      'harika', 'mükemmel', 'süper', 'fantastik', 'inanılmaz', 'muhteşem'
    ],
    negatif: [
      'üzgün', 'kızgın', 'endişeli', 'korkmuş', 'stresli', 'yorgun', 'bıkmış',
      'sıkılmış', 'kaygılı', 'panik', 'sinirli', 'öfkeli', 'hayal kırıklığı',
      'kötü', 'berbat', 'korkunç', 'dehşet', 'felaket', 'kabus'
    ],
    nötr: [
      'normal', 'sakin', 'dengeli', 'kararlı', 'sessiz', 'düşünceli', 'dikkatli'
    ]
  };

  // Tonlama kelimeleri
  static tonlamaKelimeleri = {
    resmiyet: {
      samimi: ['arkadaş', 'dost', 'sen', 'birlikte', 'beraber', 'hey', 'bak'],
      resmi: ['sayın', 'efendim', 'lütfen', 'rica ederim', 'memnuniyetle', 'önemli']
    },
    enerji: {
      düşük: ['yavaş', 'sakin', 'sessiz', 'durgun', 'sakinleş', 'dinlen'],
      yüksek: ['hızlı', 'enerjik', 'coşkulu', 'heyecanlı', 'hareketli', 'aktif']
    },
    aciliyet: {
      normal: ['zamanla', 'yavaşça', 'sakin', 'düşün', 'planla'],
      acil: ['hemen', 'şimdi', 'acil', 'hızlı', 'derhal', 'vakit kaybetme']
    }
  };

  // Mesajın duygusal tonunu analiz et
  static async analizEt(mesaj, kullaniciId, kullaniciTipi, dil = 'tr') {
    try {
      const duyguSkoru = await this._duyguSkoruHesapla(mesaj, dil);
      const anaDuygular = await this._anaDuygulariTespitEt(mesaj, dil);
      const tonAnalizi = await this._tonAnaliziYap(mesaj, dil);
      const yanitToni = await this._yanitTonuBelirle(duyguSkoru, tonAnalizi, kullaniciTipi);

      const duyguAnalizi = new DuyguAnalizi({
        kullaniciId,
        kullaniciTipi,
        mesaj,
        duyguSkoru,
        anaDuygular,
        tonAnalizi,
        dil,
        yanitToni
      });

      await duyguAnalizi.save();
      return duyguAnalizi;
    } catch (error) {
      console.error('Duygu analizi hatası:', error);
      throw error;
    }
  }

  // Duygu skorunu hesapla (-1 ile 1 arası)
  static async _duyguSkoruHesapla(mesaj, dil) {
    const lowerMesaj = mesaj.toLowerCase();
    let pozitifPuan = 0;
    let negatifPuan = 0;
    let toplamKelime = 0;

    // Pozitif kelimeleri say
    for (const kelime of this.duyguKelimeleri.pozitif) {
      const regex = new RegExp(`\\b${kelime}\\b`, 'gi');
      const eslesme = lowerMesaj.match(regex);
      if (eslesme) {
        pozitifPuan += eslesme.length;
      }
    }

    // Negatif kelimeleri say
    for (const kelime of this.duyguKelimeleri.negatif) {
      const regex = new RegExp(`\\b${kelime}\\b`, 'gi');
      const eslesme = lowerMesaj.match(regex);
      if (eslesme) {
        negatifPuan += eslesme.length;
      }
    }

    // Emoji analizi
    const pozitifEmojiler = ['😊', '😄', '😃', '😁', '😆', '😍', '🥰', '😘', '🤗', '👍', '👏', '🎉', '🎊', '✨', '🌟', '💪', '🔥', '💯', '❤️', '💖', '💕', '💗', '💓', '💝', '💞', '💟', '💌', '💋', '💍', '💎', '💐', '🌸', '🌺', '🌻', '🌼', '🌷', '🌹', '🌱', '🌲', '🌳', '🌴', '🌵', '🌾', '🌿', '☘️', '🍀', '🍁', '🍂', '🍃', '🌍', '🌎', '🌏', '🌐', '🌑', '🌒', '🌓', '🌔', '🌕', '🌖', '🌗', '🌘', '🌙', '🌚', '🌛', '🌜', '☀️', '🌝', '🌞', '⭐', '🌟', '🌠', '☁️', '⛅', '🌤️', '🌥️', '🌦️', '🌧️', '⛈️', '🌩️', '🌨️', '☃️', '⛄', '🌬️', '💨', '🌪️', '🌫️', '🌊', '💧', '💦', '☔', '☂️', '🌈', '☀️', '🌤️', '🌥️', '🌦️', '🌧️', '⛈️', '🌩️', '🌨️', '☃️', '⛄', '🌬️', '💨', '🌪️', '🌫️', '🌊', '💧', '💦', '☔', '☂️', '🌈'];
    const negatifEmojiler = ['😢', '😭', '😤', '😠', '😡', '🤬', '😈', '👿', '💀', '☠️', '💩', '🤡', '👹', '👺', '👻', '👽', '👾', '🤖', '😺', '😸', '😹', '😻', '😼', '😽', '🙀', '😿', '😾', '🙈', '🙉', '🙊', '💌', '💘', '💔', '💕', '💖', '💗', '💙', '💚', '💛', '🧡', '💜', '🖤', '💝', '💞', '💟', '❣️', '💔', '❤️‍🔥', '❤️‍🩹', '❤️', '🧡', '💛', '💚', '💙', '💜', '🤎', '🖤', '🤍', '💔', '❣️', '💕', '💞', '💓', '💗', '💖', '💘', '💝', '💟'];

    for (const emoji of pozitifEmojiler) {
      const eslesme = mesaj.match(new RegExp(emoji, 'g'));
      if (eslesme) {
        pozitifPuan += eslesme.length * 0.5;
      }
    }

    for (const emoji of negatifEmojiler) {
      const eslesme = mesaj.match(new RegExp(emoji, 'g'));
      if (eslesme) {
        negatifPuan += eslesme.length * 0.5;
      }
    }

    // Büyük harf analizi (enerji göstergesi)
    const buyukHarfSayisi = (mesaj.match(/[A-ZĞÜŞİÖÇ]/g) || []).length;
    if (buyukHarfSayisi > mesaj.length * 0.3) {
      pozitifPuan += 0.3; // Enerji yüksek
    }

    // Ünlem işareti analizi
    const unlemSayisi = (mesaj.match(/!/g) || []).length;
    pozitifPuan += unlemSayisi * 0.2;

    // Soru işareti analizi
    const soruSayisi = (mesaj.match(/\?/g) || []).length;
    if (soruSayisi > 0) {
      pozitifPuan += 0.1; // İlgi göstergesi
    }

    // Toplam puan hesaplama
    const toplamPuan = pozitifPuan + negatifPuan;
    if (toplamPuan === 0) {
      return 0; // Nötr
    }

    const skor = (pozitifPuan - negatifPuan) / toplamPuan;
    return Math.max(-1, Math.min(1, skor)); // -1 ile 1 arasında sınırla
  }

  // Ana duyguları tespit et
  static async _anaDuygulariTespitEt(mesaj, dil) {
    const lowerMesaj = mesaj.toLowerCase();
    const duygular = [];

    // Temel duygu kategorileri
    const duyguKategorileri = {
      'mutlu': ['mutlu', 'sevinçli', 'neşeli', 'gülümseme', 'gülmek', 'eğlenceli', 'harika', 'süper'],
      'üzgün': ['üzgün', 'kederli', 'mutsuz', 'ağlamak', 'hüzün', 'keder', 'acı', 'yas'],
      'kızgın': ['kızgın', 'öfkeli', 'sinirli', 'kızmak', 'öfke', 'sinir', 'kudurmak', 'çıldırmak'],
      'endişeli': ['endişeli', 'kaygılı', 'stresli', 'panik', 'korku', 'endişe', 'kaygı', 'stres'],
      'heyecanlı': ['heyecanlı', 'coşkulu', 'enerjik', 'heyecan', 'coşku', 'enerji', 'canlı', 'dinamik'],
      'yorgun': ['yorgun', 'bitkin', 'tükenmiş', 'yorulmak', 'bitmek', 'tükenmek', 'halsiz'],
      'sakin': ['sakin', 'huzurlu', 'dingin', 'sessiz', 'huzur', 'dinginlik', 'sükunet'],
      'şaşkın': ['şaşkın', 'hayret', 'şaşırmak', 'inanılmaz', 'inanmak', 'hayret etmek']
    };

    for (const [duygu, kelimeler] of Object.entries(duyguKategorileri)) {
      for (const kelime of kelimeler) {
        if (lowerMesaj.includes(kelime)) {
          if (!duygular.includes(duygu)) {
            duygular.push(duygu);
          }
          break;
        }
      }
    }

    return duygular.length > 0 ? duygular : ['nötr'];
  }

  // Ton analizi yap
  static async _tonAnaliziYap(mesaj, dil) {
    const lowerMesaj = mesaj.toLowerCase();
    
    // Resmiyet analizi
    let resmiyetSkoru = 0.5; // Varsayılan orta
    for (const kelime of this.tonlamaKelimeleri.resmiyet.samimi) {
      if (lowerMesaj.includes(kelime)) {
        resmiyetSkoru -= 0.2;
      }
    }
    for (const kelime of this.tonlamaKelimeleri.resmiyet.resmi) {
      if (lowerMesaj.includes(kelime)) {
        resmiyetSkoru += 0.2;
      }
    }

    // Enerji analizi
    let enerjiSkoru = 0.5; // Varsayılan orta
    for (const kelime of this.tonlamaKelimeleri.enerji.düşük) {
      if (lowerMesaj.includes(kelime)) {
        enerjiSkoru -= 0.2;
      }
    }
    for (const kelime of this.tonlamaKelimeleri.enerji.yüksek) {
      if (lowerMesaj.includes(kelime)) {
        enerjiSkoru += 0.2;
      }
    }

    // Aciliyet analizi
    let aciliyetSkoru = 0.3; // Varsayılan düşük
    for (const kelime of this.tonlamaKelimeleri.aciliyet.acil) {
      if (lowerMesaj.includes(kelime)) {
        aciliyetSkoru += 0.3;
      }
    }

    // Büyük harf ve ünlem işareti analizi
    const buyukHarfOrani = (mesaj.match(/[A-ZĞÜŞİÖÇ]/g) || []).length / mesaj.length;
    const unlemSayisi = (mesaj.match(/!/g) || []).length;
    
    if (buyukHarfOrani > 0.3) {
      enerjiSkoru += 0.2;
      aciliyetSkoru += 0.1;
    }
    
    if (unlemSayisi > 0) {
      enerjiSkoru += unlemSayisi * 0.1;
      aciliyetSkoru += unlemSayisi * 0.05;
    }

    return {
      resmiyet: Math.max(0, Math.min(1, resmiyetSkoru)),
      enerji: Math.max(0, Math.min(1, enerjiSkoru)),
      aciliyet: Math.max(0, Math.min(1, aciliyetSkoru))
    };
  }

  // Yanıt tonunu belirle
  static async _yanitTonuBelirle(duyguSkoru, tonAnalizi, kullaniciTipi) {
    // Çocuklar için daha samimi ve motivasyonel
    if (kullaniciTipi === 'child') {
      if (duyguSkoru < -0.3) {
        return 'destekleyici'; // Üzgün/kızgın çocuk için destek
      } else if (duyguSkoru < 0.2) {
        return 'motivasyonel'; // Nötr çocuk için motivasyon
      } else {
        return 'samimi'; // Mutlu çocuk için samimi
      }
    }
    
    // Veliler için daha profesyonel
    if (kullaniciTipi === 'parent') {
      if (tonAnalizi.aciliyet > 0.6) {
        return 'profesyonel'; // Acil durumlar için profesyonel
      } else if (duyguSkoru < -0.2) {
        return 'destekleyici'; // Stresli veli için destek
      } else {
        return 'eğitici'; // Normal durumlar için eğitici
      }
    }

    return 'samimi'; // Varsayılan
  }

  // Yanıtı duygu analizine göre ayarla
  static async yanitiAyarla(yanit, duyguAnalizi) {
    if (!duyguAnalizi) return yanit;

    let ayarlanmisYanit = yanit;

    // Duygu skoruna göre emoji ekle
    if (duyguAnalizi.duyguSkoru > 0.5) {
      ayarlanmisYanit = `😊 ${ayarlanmisYanit}`;
    } else if (duyguAnalizi.duyguSkoru < -0.5) {
      ayarlanmisYanit = `🤗 ${ayarlanmisYanit}`;
    }

    // Ton analizine göre cümle yapısını ayarla
    if (duyguAnalizi.tonAnalizi.enerji > 0.7) {
      // Yüksek enerji için daha coşkulu
      ayarlanmisYanit = ayarlanmisYanit.replace(/\./g, '!');
    } else if (duyguAnalizi.tonAnalizi.enerji < 0.3) {
      // Düşük enerji için daha sakin
      ayarlanmisYanit = ayarlanmisYanit.replace(/!/g, '.');
    }

    // Resmiyet seviyesine göre hitap şeklini ayarla
    if (duyguAnalizi.tonAnalizi.resmiyet > 0.7) {
      // Resmi ton için daha formal
      ayarlanmisYanit = ayarlanmisYanit.replace(/sen/g, 'siz');
      ayarlanmisYanit = ayarlanmisYanit.replace(/Sen/g, 'Siz');
    }

    // Aciliyet seviyesine göre hız belirteci ekle
    if (duyguAnalizi.tonAnalizi.aciliyet > 0.7) {
      ayarlanmisYanit = `Hemen: ${ayarlanmisYanit}`;
    }

    return ayarlanmisYanit;
  }

  // Kullanıcının duygu geçmişini getir
  static async kullaniciDuyguGecmisi(kullaniciId, kullaniciTipi, gunSayisi = 7) {
    try {
      const baslangicTarihi = new Date();
      baslangicTarihi.setDate(baslangicTarihi.getDate() - gunSayisi);

      const gecmis = await DuyguAnalizi.find({
        kullaniciId,
        kullaniciTipi,
        tarih: { $gte: baslangicTarihi }
      }).sort({ tarih: -1 });

      return gecmis;
    } catch (error) {
      console.error('Duygu geçmişi getirme hatası:', error);
      return [];
    }
  }

  // Duygu istatistiklerini getir
  static async duyguIstatistikleri(kullaniciId, kullaniciTipi, gunSayisi = 30) {
    try {
      const baslangicTarihi = new Date();
      baslangicTarihi.setDate(baslangicTarihi.getDate() - gunSayisi);

      const analizler = await DuyguAnalizi.find({
        kullaniciId,
        kullaniciTipi,
        tarih: { $gte: baslangicTarihi }
      });

      const istatistikler = {
        toplamMesaj: analizler.length,
        ortalamaDuyguSkoru: 0,
        duyguKategorileri: {},
        enCokKullanilanDuygular: {},
        tonAnaliziOrtalamalari: {
          resmiyet: 0,
          enerji: 0,
          aciliyet: 0
        }
      };

      if (analizler.length > 0) {
        // Ortalama duygu skoru
        const toplamSkor = analizler.reduce((sum, analiz) => sum + analiz.duyguSkoru, 0);
        istatistikler.ortalamaDuyguSkoru = toplamSkor / analizler.length;

        // Duygu kategorileri
        analizler.forEach(analiz => {
          istatistikler.duyguKategorileri[analiz.duyguKategori] = 
            (istatistikler.duyguKategorileri[analiz.duyguKategori] || 0) + 1;
        });

        // En çok kullanılan duygular
        analizler.forEach(analiz => {
          analiz.anaDuygular.forEach(duygu => {
            istatistikler.enCokKullanilanDuygular[duygu] = 
              (istatistikler.enCokKullanilanDuygular[duygu] || 0) + 1;
          });
        });

        // Ton analizi ortalamaları
        const toplamResmiyet = analizler.reduce((sum, analiz) => sum + analiz.tonAnalizi.resmiyet, 0);
        const toplamEnerji = analizler.reduce((sum, analiz) => sum + analiz.tonAnalizi.enerji, 0);
        const toplamAciliyet = analizler.reduce((sum, analiz) => sum + analiz.tonAnalizi.aciliyet, 0);

        istatistikler.tonAnaliziOrtalamalari = {
          resmiyet: toplamResmiyet / analizler.length,
          enerji: toplamEnerji / analizler.length,
          aciliyet: toplamAciliyet / analizler.length
        };
      }

      return istatistikler;
    } catch (error) {
      console.error('Duygu istatistikleri hatası:', error);
      return null;
    }
  }
}

export default DuyguAnaliziService; 